Mathlib/Analysis/Normed/Group/SeparationQuotient.lean

```diff
@@ -0,0 +1,205 @@
+`‖x‖ = ‖mk x‖` using the lift.
+
+-/
+
+
+noncomputable section
+
+open SeparationQuotient Metric Set Topology NNReal
+
+variable {M N : Type*} [SeminormedAddCommGroup M] [SeminormedAddCommGroup N]
+
+namespace SeparationQuotientAddGroup
+
+/-- The null subgroup with respect to the norm. -/
+def nullSubgroup : AddSubgroup M where
+  carrier := {x : M | ‖x‖ = 0}
+  add_mem' {x y} (hx : ‖x‖ = 0) (hy : ‖y‖ = 0) := by
+    apply le_antisymm _ (norm_nonneg _)
+    refine (norm_add_le x y).trans_eq ?_
+    rw [hx, hy, add_zero]
+  zero_mem' := norm_zero
+  neg_mem' {x} (hx : ‖x‖ = 0) := by
+    simp only [mem_setOf_eq, norm_neg]
+    exact hx
+
+lemma inseparable_iff_norm_zero (x y : M) : Inseparable x y ↔ ‖x - y‖ = 0 := by
+  rw [Metric.inseparable_iff, dist_eq_norm]
+
+lemma isClosed_nullSubgroup : IsClosed (@nullSubgroup M _ : Set M) :=
+  isClosed_singleton.preimage continuous_norm
+
+instance : Nonempty (@nullSubgroup M _) := ⟨0⟩
+
+variable (x : SeparationQuotient M)
+
+variable (z : M)
+
+/-- The norm of the image of `m : M` in the quotient by the null space is zero if and only if `m`
+belongs to the null space. -/
+theorem quotient_norm_eq_zero_iff (m : M) :
+    ‖mk m‖ = 0 ↔ m ∈ nullSubgroup := by
+  rw [norm_mk]
+  exact Eq.to_iff rfl
+
+/-- If for `(m : M)` it holds that `mk m = 0`, then `m  ∈ nullSubgroup`. -/
+theorem mk_eq_zero_iff (m : M) : mk m = 0 ↔ m ∈ nullSubgroup := by
+  rw [← quotient_norm_eq_zero_iff]
+  exact Iff.symm norm_eq_zero
+
+open NormedAddGroupHom
+
+/-- The morphism from a seminormed group to the quotient by the null space. -/
+noncomputable def normedMk : NormedAddGroupHom M (SeparationQuotient M) :=
+  { mkAddGroupHom with
+    bound' := ⟨1, fun m => by simp only [ZeroHom.toFun_eq_coe, AddMonoidHom.toZeroHom_coe,
+      mkAddGroupHom_apply, norm_mk, one_mul, le_refl]⟩}
+
+/-- `mkAddGroupHom` agrees with `QuotientAddGroup.mk`. -/
+@[simp]
+theorem normedMk.apply (m : M) : normedMk m = mk m := rfl
+
+/-- `normedMk` is surjective. -/
+theorem surjective_normedMk : Function.Surjective (@normedMk M _) :=
+  surjective_quot_mk _
+
+/-- The kernel of `normedMk` is `nullSubgroup`. -/
+theorem ker_normedMk : (@normedMk M _).ker = nullSubgroup := by
+  ext _; exact mk_eq_zero_iff _
+
+/-- The operator norm of the projection is at most `1`. -/
+theorem norm_normedMk_le : ‖(@normedMk M _)‖ ≤ 1 :=
+  NormedAddGroupHom.opNorm_le_bound _ zero_le_one fun m => by simp only [normedMk.apply, norm_mk,
+    one_mul, le_refl]
+
+lemma eq_of_inseparable (f : NormedAddGroupHom M N) (hf : ∀ x ∈ nullSubgroup, f x = 0) :
+    ∀ x y, Inseparable x y → f x = f y :=
+  fun _ _ h ↦ eq_of_sub_eq_zero <| by
+    rw [← map_sub]
+    exact hf _ <| inseparable_iff_norm_zero .. |>.mp h
+
+/-- The lift of a group hom to the separation quotient as a group hom. -/
+noncomputable def liftNormedAddGroupHom (f : NormedAddGroupHom M N)
+    (hf : ∀ x ∈ nullSubgroup, f x = 0) : NormedAddGroupHom (SeparationQuotient M) N :=
```

Comment:
I'd argue this and the ones below should be written
```suggestion
    (hf : ∀ x, ‖x‖ = 0 → f x = 0) : NormedAddGroupHom (SeparationQuotient M) N :=
```
as this is shorter, and the fact that the null space is a subgroup is sort of irrelevant here.


However, if you want to justify the use of nullSubgrup here, you could do something like
```suggestion
    (hf : f.comp normedSubgroup.subtype = 0) : NormedAddGroupHom (SeparationQuotient M) N :=
```
(possibly after defining a `NormedAddGroupHom (nullSubgroup M) M` and using that in place of `subtype`)

Reply:
I see, I took the first one

Reply:
no, I agree with the first suggestion. Indeed, as I mention in #16707, I'm still not yet convinced we even need `nullSubgroup` et al.

Reply:
I removed `nullSubgroup` from this PR